### DISCUSSION TITLE
Add mod search option

### DISCUSSION THREAD

**Discussion by Cootz** [ppy/osu#22893] [Ideas] [answered]
There are already plenty of mods you can play with and it's quite confusing to find them in mod selection. Would be nice to have a search option for this purpose.

**Comment by Axilotl17** [comment]
sounds bloated. a search bar would not help; it would just make everything more complex.

**Comment by peppy** [comment, marked as answer]
This is planned.

**Reply by Cootz** [reply]
Can I help developing this?

**Reply by peppy** [reply]
@bdach already made a start on this, so at very least you'd want to check with him first.

**Reply by bdach** [reply]
I have a work in progress branch that I could not bring to a state I would consider acceptable for several months now due to several issues in trying to fit `SearchContainer` into the mods overlay. If you _really_ wish to try then I'm not going to stop you, but it's just something you should keep in mind as various deeper-level changes may be required to make it work well.

**Reply by bdach** [reply]
Do let me know if you're going to try this though, thanks.

**Reply by Cootz** [reply]
@bdach I'm going to try this. I didn't find any branch with search prototype on your osu fork so I'll create it from scratch

**Reply by bdach** [reply]
@Cootz if it helps see https://github.com/ppy/osu/compare/master...bdach:osu:mod-select-filtering-old. in particular see https://github.com/ppy/osu/commit/2908cecea2b24f98f8e7c9afadcc30fbdbbacf47 which describes the blocking issues. the `ModColumn` one is particularly debilitating.

i believe `IFilterable` is `IConditionalFilterable` now, but the rest should somewhat work. excluding the blockers, obviously.

i should also emphasise that opening PRs without fixing the blockers mentioned is useless.

**Reply by Cootz** [reply]
@bdach Thank you for sharing this code, it helped a lot. I fixed `ModSelectColumn` disappearing from `ModSelectOverlay` see [`master...Cootz:osu:add-mod-search-option`](https://github.com/ppy/osu/compare/master...Cootz:osu:add-mod-search-option). I would like you to take a look at it and tell if this workaround can be considered as acceptable?

If no can we consider adding a public `SearchContainer.Filter` method to `osu.framework` to manually invoke filtering when we need it?

**Reply by bdach** [reply]
What you linked is not a good workaround. It's pretty much as ad-hoc as you can get.

I did consider adding a public method to force refiltering but I wasn't sure how it would be received in review. It's still pretty bad but not *horribly* bad. Try that first I guess.
